File: ShareX_windows/ShareX_windows/Helpers/MainSocket.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Net.Sockets;
using System.Text;
using System.Threading;
using System.Threading.Tasks;

namespace ShareX_windows.Helpers
{


    public enum SocketMode
    {
        SendingFile,
        ReceivingFile,
        SendPhotos,
        ReceivePhotos,
        SendMessage,
        ReciveMessage,
        GetImage,
        Any
    }
    public class MainSocket
    {
        private Action receiveFileTask;
        private Action receiveNotificationTask;

        private static MainSocket instance = null;
        public static MainSocket Instance
        {
            get
            {
                if(instance == null)
                {
                    instance = new MainSocket();
                }
                return instance;
            }
        }
        private int mainSocketPort = 6774;
        private SocketHelper _mainSocket;
        private MainSocket()
        {
            _mainSocket = new(port: mainSocketPort);
        }

        public async Task<bool> connect()
        {
            Debug.WriteLine($"socket is waiting");
            await _mainSocket.hostServer();
            Debug.WriteLine($"socket is connected");
            ReceiveThread();
            return true;

        }

       

        public async void ReceiveThread()
        {
           
            SocketMode currentMode = SocketMode.Any;

            while (true)
            {
                try
                {
                    var rcvMsg = await _mainSocket.receiveMsg();
                    int rcvInt = -1;

                    if (rcvMsg != "" && rcvMsg != null)
                    {
                        if (int.TryParse(rcvMsg, out rcvInt))
                        {
                            currentMode = (SocketMode)rcvInt;
                            Debug.WriteLine($"socket mode is {currentMode}");

                            if (!Enum.IsDefined(typeof(SocketMode), currentMode) && !currentMode.ToString().Contains(","))
                            {
                                //throw new InvalidOperationException(
                                //    $"{recMsg} is not an underlying value of the YourEnum enumeration."
                                //);
                                continue;
                            }
                            switch (currentMode)
                            {
                                case SocketMode.SendingFile:
                                    if(receiveFileTask != null)
                                        receiveFileTask();
                                    break;
                                case SocketMode.ReceivingFile:
                                    break;
                                case SocketMode.SendPhotos:
                                    break;
                                case SocketMode.ReciveMessage:
                                    break;
                                case SocketMode.SendMessage:
                                    break;
                                case SocketMode.Any:
                                    break;
                                default:
                                    break;
                            }
                        }
                        else
                        {
                            //Debug.WriteLine($"rcv msg is {rcvMsg}");
                        }
                    }
                }
                catch (SocketException e)
                {
                    Debug.WriteLine("mainSocket receive Therad :" + e.Message);
                    await _mainSocket.hostServer();
                }
            }
        }

        public async void sendMsg(SocketMode mode)
        {
            Debug.WriteLine($"socket name {mode.ToString()}");
            var sendStr = mode.ToString();
            await _mainSocket.send_msg(sendStr);
        }

        public void receiveFileCallback(Action task)
        {   
            receiveFileTask =  task;
        }
        public void receiveNotificationCallback(Action task)
        {   
            receiveNotificationTask =  task;
        }
    }
}

File: ShareX_android/app/src/main/java/com/example/sharex/helpers/MainSocket.kt
package com.example.sharex.helpers

import android.util.Log
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

enum class SocketMode {
    SendingFile,
    ReceivingFile,
    SendPhotos,
    ReceivePhotos,
    SendMessage,
    ReceiveMessage,
    GetImage,
    Notification,
    Any
}

object MainSocket {

    private const val port = 6774

    private var receiveFile: (() -> Unit)? = null
    private lateinit var socket: SocketHelper

    public var isConnected = false
    public fun connectMainSocket(ip: String): Boolean {
        socket = SocketHelper(port, "Main_server")
        CoroutineScope(Dispatchers.IO).launch {

            while (!isConnected) {
                Log.d("mainSocket1", "mainSocket waiting")
                isConnected = socket.connectToServer(ip)
                if (isConnected)
                    Log.d("mainSocket1", "mainSocket connected")
                else
                    Log.d("mainSocket1", "mainSocket not connected")
            }

            receiveThread()
        }
        return isConnected
    }

    private suspend fun receiveThread() {
        var currentMode = SocketMode.Any;
        while (true) {
            if(isConnected)
            {
                try {
                    val rcvMsg = socket.receiveMsg();
                    if (rcvMsg.isEmpty())
                        continue
                    Log.d("rcvmsg", "rcvmsg is $rcvMsg")
                    var rcvInt: Int? = -1;
                    if (rcvMsg != "") {
                        if (rcvInt != null) {
                            rcvInt = rcvMsg.toIntOrNull()
                            currentMode = SocketMode.valueOf(rcvMsg);
                            Log.d("mainSocket1", "$currentMode is not in socket mode ENUM 111")

                            when (currentMode) {
                                SocketMode.SendingFile -> {
                                    if (receiveFile != null) {
                                        Log.d(
                                            "mainSocket1",
                                            "$currentMode is not in socket mode ENUM 111"
                                        )
                                        receiveFile?.invoke()
                                    }
                                }
                                else -> {
                                    Log.d("mainSocket1", "$currentMode is not in socket mode ENUM")
                                }
                            }
                        } else {
                            //Debug.WriteLine($"rcv msg is {rcvMsg}");
                        }
                    }
                } catch (e: Exception) {
//                    Log.d("rcvmsg", "rcvmsg is error ${e.message}")
                }
            }
        }

    }

    public fun sendMsg(mode: SocketMode) {
        CoroutineScope(Dispatchers.IO).launch {
            if(isConnected)
            {
                Log.d("mainSocket1", "message sent from main socket")
                socket.sendMsg(mode.ordinal.toString())
            }
        }
    }

    public fun disconnect() {
        CoroutineScope(Dispatchers.IO).launch {
            socket.disconnect()
        }
    }

    public fun receiveCallback(receiveFun: () -> Unit) {
        receiveFile = receiveFun
    }
}